qiskit_experiments/framework/backend_timing_mixin.py

```diff
@@ -0,0 +1,290 @@
+      particular, since the ``x`` gates should have durations that are
+      multiples of ``acquire_alignment`` (because ``granularity`` usually is),
+      the ``measure`` start will occur at a time consistent with
+      ``acquire_alignment`` when ``delay + delay2`` is a multiple of
+      ``acquire_alignment``. Note that in the case of IBM Quantum backends,
+      when ``acquire_alignment`` is not satisfied, there is no error reported
+      by Qiskit or by the backend. Instead the measurement pulse is misaligned
+      relative to the start of the signal acquisition, resulting in an
+      incorrect phase and often an incorrect state discrimination.
+
+    To help avoid these problems, :class:`.BackendTiming` provides methods for
+    calculating pulse and delay durations in samples and seconds, for a given
+    input duration in seconds. If these values are used for all durations in a
+    circuit, the alignment constraints should always be satisfied.
+
+    .. note::
+
+        For delay duration, the least common multiple of ``pulse_alignment``
+        and ``acquire_alignment`` is used as the granularity. Thus, in the
+        ``acquire_alignment`` example above, ``delay`` and ``delay2`` are each
+        a multiple of ``acquire_alignment`` and so the sum always is. This
+        approach excludes some valid circuits (like each delay being half of
+        ``acquire_alignment``) but has the benefit of always being valid
+        without detailed analysis of the full circuit.
+
+    TODO: add example usage
+    TODO: expose to framework and documentation
+
+    .. note::
+
+        The methods in this class assume that the ``backend`` attribute is
+        constant. Methods should not call methods of this class before and
+        after modifying the ``backend`` attribute and expect consistent
+        results. In particular, when the backend is not set, no ``dt`` value is
+        available and :meth:`.BackendTiming.circuit_delay` will return the
+        delay time in seconds whereas once the backend is set the value
+        returned will be in units of ``dt`` rounded according to the alignment
+        constraints.
+    """
+
+    def __init__(self, experiment: BaseExperiment):
+        """Initialize backend timing object
+
+        Args:
+            experiment: the experiment to provide timing help for
+        """
+        self.experiment = experiment
+
+        self._backend: Union[Backend, None] = None
+        self._backend_data_cached: Union[BackendData, None] = None
+
+    @property
+    def _backend_data(self) -> BackendData:
+        """Backend data associated with experiment
+
+        Returns:
+            The BackendData object associated with the experiment
+
+        Raises:
+            QiskitError: if the backend is not set on the experiment
+        """
+        if self.experiment.backend is None:
+            raise QiskitError("Backend not set on experiment!")
+
+        if self.experiment.backend != self._backend:
```

Comment:
Backend object doesn't explicitly define `__eq__`. How does this work? I guess this covers the case that an experiment is instantiated with backend A and run with backend B. Perhaps we can just compare backend.name?
https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/backend.py#L268

Reply:
When `__eq__` is not defined, it falls back to `is` which is fine here. This comparison is just for caching, so as long as it does not incorrectly evaluate to `True` it is okay if it occasionally unnecessarily evaluates to `False`.

Really, this caching might have been premature optimization. I chose the experiment as the input to BackendTiming because I originally thought of `BackendTiming` as a mixin class that would have access to `self`. The backend_data had to be guarded by the backend which could be changed if set_backend was called on the experiment. To allow for this, backend_data would have to be recreated on every use or cached in some way. Since it does a bit of work in `__init__` I thought it was better to cache it.

However, now that BackendTiming is a standalone object rather than a mixin it could take a backend directly. Then there is no need for it to be a long-lived object that could have the backend changed underneath it. If the class takes backend as an input directly it can assume that backend_data never changes. I will try this since it makes the class simpler.

Reply:
See d4838817